src/mbgl/renderer/image_manager.cpp

```diff
@@ -137,48 +147,57 @@ void ImageManager::notifyIfMissingImageAdded() {
     }
 }
 
-void ImageManager::reduceMemoryUse() {
-    for (auto it = requestedImages.cbegin(); it != requestedImages.cend();) {
-        if (it->second.empty() && images.find(it->first) != images.end()) {
-            images.erase(it->first);
-            removePattern(it->first);
-            it = requestedImages.erase(it);
-        } else {
-            ++it;
+void ImageManager::reduceMemoryUse() const {
+    std::vector<std::string> unusedIDs;
+    unusedIDs.reserve(requestedImages.size());
+
+    for (const auto& kv : requestedImages) {
+        if (kv.second.empty() && images.find(kv.first) != images.end()) {
+            unusedIDs.push_back(kv.first);
         }
     }
+
+    if (!unusedIDs.empty()) {
+        observer->onRemoveUnusedStyleImages(std::move(unusedIDs));
+    }
+}
+
+void ImageManager::checkCacheSizeReduceMemoryUse() const {
+    if (requestedImagesCacheSize > util::DEFAULT_ON_DEMAND_IMAGES_CACHE_SIZE) {
+        reduceMemoryUse();
+    }
 }
 
 void ImageManager::checkMissingAndNotify(ImageRequestor& requestor, const ImageRequestPair& pair) {
-    unsigned int missing = 0;
+    std::vector<std::string> missingImages;
+    missingImages.reserve(pair.first.size());
     for (const auto& dependency : pair.first) {
-        auto it = images.find(dependency.first);
-        if (it == images.end()) {
-            missing++;
-            requestedImages[dependency.first].emplace(&requestor);
+        if (images.find(dependency.first) == images.end()) {
+            missingImages.push_back(dependency.first);
         }
     }
 
-    if (missing > 0) {
+    if (!missingImages.empty()) {
         ImageRequestor* requestorPtr = &requestor;
-
-        missingImageRequestors.emplace(requestorPtr, MissingImageRequestPair { std::move(pair), missing });
-
+        missingImageRequestors.emplace(requestorPtr, MissingImageRequestPair { pair, missingImages.size() });
```

Comment:
`pair` is not used further, right? so should be OK to move it

Reply:
pair const ref, move would have no effect.

Reply:
`std::move` will make it rvalue, so it will take effect on constructing  MissingImageRequestPair -> ImageRequestPair -> ImageDependencies (aka unordered_map)

Reply:
`std::move` for `const ImageRequestPair&` will result in `const ImageRequestPair&&`, std::pair would be constructed using `const T&` overload, right?

Reply:
You're right! Sorry for the hassle